File: backend/app/schemas/notification.py
"""Pydantic response contracts for the notification inbox + catalog APIs.

Response schemas inherit ``UTCModel`` so datetimes serialize as UTC ISO-8601 with a
trailing ``Z`` (store UTC, serve UTC, display Central).
"""

from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.core.pagination import PaginationMeta
from app.schemas.base import UTCModel


class NotificationResponse(UTCModel):
    id: int
    event_key: str
    severity: str
    title: str
    body: Optional[str] = None
    link: Optional[str] = None
    related_type: Optional[str] = None
    related_id: Optional[int] = None
    is_read: bool
    read_at: Optional[datetime] = None
    created_at: datetime


class NotificationListResponse(UTCModel):
    items: List[NotificationResponse]
    pagination: PaginationMeta


class UnreadCountResponse(BaseModel):
    count: int


class MarkAllReadResponse(BaseModel):
    updated: int


class CatalogEntryResponse(BaseModel):
    event_key: str
    label: str
    description: str
    category: str
    severity: str
    default_channels: List[str]
    mandatory_channel: Optional[str] = None
    sms_eligible: bool


# ---------------------------------------------------------------------------
# Self-service notification preferences
#
# PR 4 slice: the SMS channel only. The PERSISTED JSON keeps the full
# ``{in_app, email, sms, digest}`` shape per event so PR 3's complete matrix extends
# this without a migration; the REQUEST model accepts only ``sms`` and forbids extra
# keys, so a PR-3-shaped payload fails loudly (422) instead of silently dropping the
# channels this endpoint does not yet own.
# ---------------------------------------------------------------------------


class NotificationChannelUpdate(BaseModel):
    """Per-event channel changes. PR 4 scope: ``sms`` only."""

    model_config = ConfigDict(extra="forbid")

    sms: bool = Field(..., description="Deliver this event over SMS (SMS-eligible events only)")


class NotificationPreferencesUpdate(BaseModel):
    """Body of ``PUT /users/me/notification-preferences`` (self-scoped)."""

    model_config = ConfigDict(extra="forbid")

    preferences: Dict[str, NotificationChannelUpdate] = Field(
        ...,
        max_length=200,
        description="Map of catalog event_key -> channel changes",
    )


class NotificationPreferencesResponse(BaseModel):
    """Effective notification preferences for the current user.

    ``preferences`` is the RESOLVED per-event channel map the dispatcher would apply
    right now (catalog defaults where the user has saved nothing, plus any mandatory
    channel forced on) — not the raw stored row, so the UI can never disagree with
    what actually gets sent.
    """

    preferences: Dict[str, Dict[str, bool]]
    has_saved_preferences: bool
    phone: Optional[str] = None
    sms_egress_enabled: bool = False
    sms_configured: bool = False


class TestSMSResponse(BaseModel):
    """Result of ``POST /users/me/test-sms``."""

    status: str
    sid: Optional[str] = None
    provider_status: Optional[str] = None
    detail: str

File: frontend/src/types/notification.ts
/**
 * Notification inbox + catalog types.
 *
 * These mirror the backend response contracts in
 * `backend/app/schemas/notification.py` (NotificationResponse,
 * NotificationListResponse, UnreadCountResponse, CatalogEntryResponse) and the
 * shared `PaginationMeta` from `backend/app/core/pagination.py`. Datetimes arrive
 * as UTC ISO-8601 with a trailing `Z` and are rendered in shop-local Central time.
 */

export type NotificationSeverity = 'info' | 'warning' | 'critical';

/** One in-app inbox row for the current user (matches `NotificationResponse`). */
export interface NotificationItem {
  id: number;
  event_key: string;
  severity: string; // one of NotificationSeverity, kept loose for forward-compat
  title: string;
  body?: string | null;
  /** Relative in-app route this notification deep-links to (e.g. `/work-orders/42`). */
  link?: string | null;
  related_type?: string | null;
  related_id?: number | null;
  is_read: boolean;
  read_at?: string | null;
  created_at: string;
}

/** Offset/limit pagination metadata (matches backend `PaginationMeta`). */
export interface PaginationMeta {
  page: number;
  page_size: number;
  total_count: number;
  total_pages: number;
  has_next: boolean;
  has_previous: boolean;
}

/** Paged inbox response (matches `NotificationListResponse`). */
export interface NotificationListResponse {
  items: NotificationItem[];
  pagination: PaginationMeta;
}

/** Query params for the paged inbox endpoint. */
export interface NotificationListParams {
  page?: number;
  pageSize?: number;
  /** `true` = only unread, `false` = only read, omit = all. */
  unread?: boolean;
  category?: string;
  severity?: string;
}

/** One catalog entry driving the settings matrix (matches `CatalogEntryResponse`). */
export interface NotificationCatalogEntry {
  event_key: string;
  label: string;
  description: string;
  category: string;
  severity: string;
  default_channels: string[];
  mandatory_channel?: string | null;
  sms_eligible: boolean;
}

/** Delivery channels a per-event preference can carry. */
export type NotificationChannel = 'in_app' | 'email' | 'sms' | 'digest';

/**
 * Per-event channel flags, keyed by channel. Read from the RESOLVED map the
 * preferences endpoint returns (catalog defaults already applied and any mandatory
 * channel already forced on), so a channel key may legitimately be absent for a
 * channel the catalog does not describe.
 */
export type NotificationEventPreference = Partial<Record<NotificationChannel, boolean>>;

/**
 * The current user's effective notification preferences — matches
 * `NotificationPreferencesResponse` (`GET`/`PUT /users/me/notification-preferences`).
 *
 * `preferences` is the resolved per-event channel map the dispatcher would apply
 * right now, NOT the raw stored row, so the UI can never disagree with what
 * actually gets sent. The three SMS flags exist so the UI can explain exactly why
 * an SMS toggle would currently be inert (no number / company egress off / provider
 * not configured). `sms_configured` is a plain boolean — no provider credential is
 * ever returned to the client.
 */
export interface NotificationPreferences {
  preferences: Record<string, NotificationEventPreference>;
  has_saved_preferences: boolean;
  phone?: string | null;
  sms_egress_enabled: boolean;
  sms_configured: boolean;
}

/**
 * Body of `PUT /users/me/notification-preferences`. The backend declares
 * `extra="forbid"` and PR 4 owns the SMS channel only, so each entry carries
 * exactly `{ sms }` — send only the events you are changing.
 */
export interface NotificationPreferencesUpdate {
  preferences: Record<string, { sms: boolean }>;
}

/**
 * Result of the self-service "send test SMS" action (`TestSMSResponse`). Only
 * `detail` is meant for display; `sid` / `provider_status` are provider
 * bookkeeping and are deliberately NOT surfaced in the UI.
 */
export interface TestSmsResponse {
  status: string;
  sid?: string | null;
  provider_status?: string | null;
  detail: string;
}
